### DISCUSSION TITLE
Y2 Step Pin Not Working

### DISCUSSION THREAD

**Discussion by bear-fighter** [grblHAL/RP2040#50] [Q&A] [answered]
I built a copy of the RP2040 driver using the picobob_map.h file.  In my_machine.h I turned on Y_AUTO_SQUARE but when I issue a Y-Axis jog command from cncjs I see the Y1 step signal but not the Y2 step signal.  As far as I'm aware, the following code in picobob_map.h tells me it should be GPIO3 that I'm looking for the step signal on.  I checked the direction pin for Y2 (GPIO12) and I am seeing that change when jogging back and forth but I'm still not seeing GPIO3 change.

Any help would be appreciated.

```
// Define ganged axis or A axis step pulse and step direction output pins.
#if N_ABC_MOTORS > 0
#define M3_AVAILABLE
#define M3_STEP_PIN           3 //not the Pico pin.
#define M3_DIRECTION_PIN      12
#define M3_LIMIT_PIN          5
#endif
```

**Comment by andrewmarles** [comment, marked as answer]
As noted in the map file there, the '3' does not reference the pico pin.  It references the pin for the PIO state machine that is used to generate steps.

https://github.com/grblHAL/RP2040/blob/833128efceecbf4e2aead48c8b13957a1e81dd3f/picobob_map.h#L33

That line shows that the PIO base pin is 17, and it uses n_axis number of consecutive pins (plus additional ganged axes) so in your case the Y2 step signal is on GPIO20.

The direction pin is not driven by PIO so it is referenced directly.

**Reply by bear-fighter** [reply]
That's exactly it!  I saw the not about not being the pico pin but it didn't say what pin it actually was.  I was probing with my oscope and found that it was GPIO20 which makes sense given that it's X (GPIO17), Y (GPIO18), Z (GPIO19), and Y2 (GPIO20).  I've been working with the rp2040 for awhile now but never went much into PIO.  Reading through the source code it's hard to track down that implementation if you've never worked with the PIO subsystem before.  Checked through the rp2040 datasheet and read up on the PIO and it makes more sense now.  Thanks!
